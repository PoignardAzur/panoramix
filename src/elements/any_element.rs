use crate::ctx::{ProcessEventCtx, ReconcileCtx};
use crate::element_tree::{Element, VirtualDom};
use crate::glue::GlobalEventCx;
use crate::metadata::NoState;
use crate::widgets::WidgetSeqBox;

use std::any::Any;
use std::fmt::Debug;

// --- STATE ---

pub trait AnyState: Any + Debug {
    fn as_any(&self) -> &dyn Any;
    fn as_mut_any(&mut self) -> &mut dyn Any;

    fn print_type(&self) {
        println!("{:#?}", std::any::type_name::<Self>());
    }

    fn dyn_clone(&self) -> Box<dyn AnyState>;
    fn dyn_eq(&self, other: &Box<dyn AnyState>) -> bool;
}

impl<T> AnyState for T
where
    T: Clone + Default + Debug + PartialEq + 'static,
{
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_mut_any(&mut self) -> &mut dyn Any {
        self
    }

    fn dyn_clone(&self) -> Box<dyn AnyState> {
        Box::new(self.clone())
    }

    fn dyn_eq(&self, other: &Box<dyn AnyState>) -> bool {
        if let Some(other) = other.as_ref().as_any().downcast_ref::<Self>() {
            other.eq(self)
        } else {
            false
        }
    }
}

pub struct AnyStateBox {
    value: Box<dyn AnyState>,
}

impl AnyStateBox {
    pub fn new(state: impl AnyState) -> Self {
        AnyStateBox {
            value: Box::new(state),
        }
    }
}

impl Clone for AnyStateBox {
    fn clone(&self) -> Self {
        AnyStateBox {
            value: self.value.dyn_clone(),
        }
    }
}

impl Debug for AnyStateBox {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.value.fmt(f)
    }
}

impl PartialEq for AnyStateBox {
    fn eq(&self, other: &Self) -> bool {
        self.value.dyn_eq(&other.value)
    }
}

// --- ELEMENT ---

#[derive(Default, Clone, PartialEq, Eq, Hash)]
struct ErasedElement<Child: Element> {
    child: Option<Child>,
}

impl<Child: Element> Debug for ErasedElement<Child> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.as_ref().unwrap().fmt(f)
    }
}

trait AnyElement: Any + Debug {
    type Event;

    fn print_type(&self) {
        println!("{:#?}", std::any::type_name::<Self>());
    }

    fn dyn_clone(&self) -> Box<dyn AnyElement<Event = Self::Event>>;

    fn build(
        &mut self,
        prev_state: Option<AnyStateBox>,
    ) -> (
        Box<dyn AnyVirtualDom<Event = Self::Event>>,
        Option<AnyStateBox>,
    );
}

impl<Child: Element + 'static> AnyElement for ErasedElement<Child> {
    type Event = Child::Event;

    fn dyn_clone(&self) -> Box<dyn AnyElement<Event = Self::Event>> {
        Box::new(self.clone())
    }

    fn build(
        &mut self,
        prev_state: Option<AnyStateBox>,
    ) -> (
        Box<dyn AnyVirtualDom<Event = Self::Event>>,
        Option<AnyStateBox>,
    ) {
        let child = self.child.take().unwrap();

        let prev_state = if let Some(mut prev_state) = prev_state {
            std::mem::take(
                prev_state
                    .value
                    .as_mut_any()
                    .downcast_mut::<Child::AggregateChildrenState>()
                    .unwrap(),
            )
        } else {
            Default::default()
        };

        let (output, state) = child.build(prev_state);

        (
            Box::new(ErasedVirtualDom { child: output }),
            Some(AnyStateBox {
                value: Box::new(state),
            }),
        )
    }
}

// -

pub struct ElementBox<Event> {
    child: Box<dyn AnyElement<Event = Event>>,
}

impl<Event> ElementBox<Event> {
    pub fn new(child: impl Element<Event = Event> + 'static) -> Self {
        ElementBox {
            child: Box::new(ErasedElement { child: Some(child) }),
        }
    }
}

impl<Event> Debug for ElementBox<Event> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.fmt(f)
    }
}

impl<Event> Clone for ElementBox<Event> {
    fn clone(&self) -> Self {
        ElementBox {
            child: self.child.dyn_clone(),
        }
    }
}

impl<Event> Element for ElementBox<Event> {
    type Event = Event;
    type ComponentState = NoState;
    type AggregateChildrenState = Option<AnyStateBox>;
    type BuildOutput = VirtualDomBox<Event>;

    fn build(self, prev_state: Option<AnyStateBox>) -> (VirtualDomBox<Event>, Option<AnyStateBox>) {
        let mut child = self.child;
        let (output, state) = child.build(prev_state);

        (VirtualDomBox { child: output }, state)
    }
}

// --- VIRTUAL_DOM ---

#[derive(Default, Clone, PartialEq, Eq, Hash)]
pub struct ErasedVirtualDom<Child: VirtualDom> {
    child: Child,
}

impl<Child: VirtualDom> Debug for ErasedVirtualDom<Child> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.fmt(f)
    }
}

pub trait AnyVirtualDom: Any + Debug {
    type Event;

    fn as_any(&self) -> &dyn Any;

    fn print_type(&self) {
        println!("{:#?}", std::any::type_name::<Self>());
    }

    fn init_tree(&self) -> WidgetSeqBox;

    fn reconcile(
        &self,
        other: &Box<dyn AnyVirtualDom<Event = Self::Event>>,
        widget_seq: &mut WidgetSeqBox,
        ctx: &mut ReconcileCtx,
    );

    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    );

    fn process_local_event(
        &self,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event>;
}

impl<Child: VirtualDom + 'static> AnyVirtualDom for ErasedVirtualDom<Child> {
    type Event = Child::Event;

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn init_tree(&self) -> WidgetSeqBox {
        WidgetSeqBox {
            value: Box::new(self.child.init_tree()),
        }
    }

    fn reconcile(
        &self,
        other: &Box<dyn AnyVirtualDom<Event = Self::Event>>,
        widget_seq: &mut WidgetSeqBox,
        ctx: &mut ReconcileCtx,
    ) {
        let other = other.as_any().downcast_ref::<Self>().unwrap();
        let widget_seq = widget_seq
            .value
            .as_mut_any()
            .downcast_mut::<Child::TargetWidgetSeq>()
            .unwrap();
        self.child.reconcile(&other.child, widget_seq, ctx);
    }

    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) {
        let children_state = children_state
            .as_mut()
            .unwrap()
            .value
            .as_mut_any()
            .downcast_mut::<Child::AggregateChildrenState>()
            .unwrap();
        let widget_seq = widget_seq
            .value
            .as_mut_any()
            .downcast_mut::<Child::TargetWidgetSeq>()
            .unwrap();
        self.child
            .process_event(comp_ctx, children_state, widget_seq, cx)
    }

    fn process_local_event(
        &self,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        let children_state = children_state
            .as_mut()
            .unwrap()
            .value
            .as_mut_any()
            .downcast_mut::<Child::AggregateChildrenState>()
            .unwrap();
        let widget_seq = widget_seq
            .value
            .as_mut_any()
            .downcast_mut::<Child::TargetWidgetSeq>()
            .unwrap();
        self.child
            .process_local_event(children_state, widget_seq, cx)
    }
}

// -

pub struct VirtualDomBox<Event> {
    child: Box<dyn AnyVirtualDom<Event = Event>>,
}

impl<Event> VirtualDomBox<Event> {
    pub fn new(child: impl VirtualDom<Event = Event> + 'static) -> Self {
        VirtualDomBox {
            child: Box::new(ErasedVirtualDom { child: child }),
        }
    }
}

impl<Event> Debug for VirtualDomBox<Event> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        self.child.fmt(f)
    }
}

impl<Event> VirtualDom for VirtualDomBox<Event> {
    // TODO - Accept any Event parent of Self::Event
    type Event = Event;
    type AggregateChildrenState = Option<AnyStateBox>;
    type TargetWidgetSeq = WidgetSeqBox;

    fn init_tree(&self) -> Self::TargetWidgetSeq {
        self.child.init_tree()
    }

    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        self.child.reconcile(&other.child, widget_seq, ctx);
    }

    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Option<AnyStateBox>,
        widget_seq: &mut WidgetSeqBox,
        cx: &mut GlobalEventCx,
    ) {
        self.child
            .process_event(comp_ctx, children_state, widget_seq, cx)
    }

    fn process_local_event(
        &self,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        self.child
            .process_local_event(children_state, widget_seq, cx)
    }
}

// --- TESTS ---

#[cfg(test)]
mod tests {
    use super::*;
    use crate::elements::{Button, Label};
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn any_state() {
        let state1 = AnyStateBox::new(42);
        let state2 = state1.clone();

        assert!(state1 == state2);
    }

    #[test]
    fn new_element() {
        let label = ElementBox::new(Label::new("Hello"));
        assert_debug_snapshot!(label);

        let (label_data, _state) = label.build(None);
        assert_debug_snapshot!(label_data);

        // TODO - check state
    }

    #[test]
    fn boxed_label_widget() {
        let label = ElementBox::new(Label::new("Hello"));

        Harness::run_test_window(label, |harness| {
            let label_state = harness.get_root_debug_state();
            assert_debug_snapshot!(label_state);

            let new_label = ElementBox::new(Label::new("World"));
            harness.update_root_element(new_label);

            let label_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(label_state_2);
        });
    }

    // FIXME - Test doesn't use box
    #[test]
    fn boxed_button_press() {
        use crate::elements::event_logger::EventLogger;
        use crate::glue::WidgetId;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let button_id = WidgetId::reserved(1);
        let button = EventLogger::new(
            event_sender,
            Button::new("Hello").with_reserved_id(button_id),
        );

        Harness::run_test_window(button, |harness| {
            assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(button_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);
        });
    }
}
