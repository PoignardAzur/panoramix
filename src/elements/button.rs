use crate::glue::{Action, GlobalEventCx, WidgetId};

use crate::element_tree::{Element, ElementExt, Metadata, NoState, VirtualDom};
use crate::flex::FlexParams;
use crate::widgets::ButtonWidget;

use crate::element_tree::ReconcileCtx;

use derivative::Derivative;
use tracing::{instrument, trace};

/// A button with a text label.
///
/// ## Events
///
/// Emits [ButtonClick] events.
#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct Button {
    pub text: String,
    pub flex: FlexParams,
    pub reserved_widget_id: Option<WidgetId>,
}

#[derive(Derivative, PartialEq)]
#[derivative(Debug(bound = ""), Default(bound = ""), Clone(bound = ""))]
pub struct ButtonData {
    pub text: String,
    pub flex: FlexParams,
    pub reserved_widget_id: Option<WidgetId>,
}

/// Event emitted when a [Button] is clicked.
///
/// Note: Might hold data like "mouse position" or "button id" future versions.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct ButtonClick;

//
// --- IMPLS

impl Button {
    /// Build a button with the given label.
    ///
    /// Use the [.on_click](Button::on_click) method to provide a closure to be called when the button is clicked.
    pub fn new(text: impl Into<String>) -> Self {
        Button {
            text: text.into(),
            flex: FlexParams {
                flex: 1.0,
                alignment: None,
            },
            reserved_widget_id: None,
        }
    }

    /// Change the way the button's size is calculated
    pub fn with_flex_params(self, flex_params: FlexParams) -> Self {
        Button {
            flex: flex_params,
            ..self
        }
    }

    // TODO - doc
    // For unit tests only
    pub fn with_reserved_id(self, widget_id: WidgetId) -> Self {
        Button {
            reserved_widget_id: Some(widget_id),
            ..self
        }
    }

    /// Provide a closure to be called when this button is clicked.
    pub fn on_click<ComponentEvent: 'static, ComponentState: 'static>(
        self,
        md: Metadata<ComponentEvent, ComponentState>,
        callback: impl Fn(&mut ComponentState, ButtonClick) + Clone,
    ) -> impl Element {
        self.on(md, callback)
    }
}

impl Element for Button {
    type Event = ButtonClick;

    type ComponentState = NoState;
    type AggregateChildrenState = ();
    type BuildOutput = ButtonData;

    #[instrument(name = "Button", skip(self, _prev_state))]
    fn build(self, _prev_state: ()) -> (ButtonData, ()) {
        (
            ButtonData {
                text: self.text,
                flex: self.flex,
                reserved_widget_id: self.reserved_widget_id,
            },
            (),
        )
    }
}

impl VirtualDom for ButtonData {
    type Event = ButtonClick;
    type AggregateChildrenState = ();
    type TargetWidgetSeq = ButtonWidget;

    #[instrument(name = "Button", skip(self))]
    fn init_tree(&self) -> ButtonWidget {
        let id = self.reserved_widget_id.unwrap_or_else(WidgetId::next);
        ButtonWidget::new(self.text.clone(), self.flex, id)
    }

    #[instrument(name = "Button", skip(self, _other, _widget, _ctx))]
    fn reconcile(&self, _other: &Self, _widget: &mut ButtonWidget, _ctx: &mut ReconcileCtx) {
        //widget.set_text(self.text.clone());
    }

    #[instrument(name = "Button", skip(self, _children_state, widget, cx))]
    fn process_local_event(
        &self,
        _children_state: &mut Self::AggregateChildrenState,
        widget: &mut ButtonWidget,
        cx: &mut GlobalEventCx,
    ) -> Option<ButtonClick> {
        // FIXME - Rework event dispatching
        let id = widget.id();
        if let Some(Action::Clicked) = cx.app_data.dequeue_action(id) {
            trace!("Processed button press");
            Some(ButtonClick)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::element_tree::assign_empty_state_type;
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn new_button() {
        let button = Button::new("Hello");
        let (button_data, ()) = button.clone().build(());

        assert_debug_snapshot!(button);
        assert_debug_snapshot!(button_data);

        assert_eq!(
            button_data,
            ButtonData {
                text: String::from("Hello"),
                flex: FlexParams {
                    flex: 1.0,
                    alignment: None,
                },
                ..Default::default()
            }
        );

        assign_empty_state_type(&button);
    }

    #[test]
    fn button_widget() {
        // TODO - We use Tuple! because RootWidget currently wants a root element with no event
        use crate::Tuple;
        let button = Tuple!(Button::new("Hello"));

        Harness::run_test_window(button, |harness| {
            let button_state = harness.get_root_debug_state();
            assert_debug_snapshot!(button_state);

            // TODO - Test reconcile() method (currently doesn't work)
        });
    }

    #[test]
    fn button_press() {
        use crate::elements::event_logger::EventLogger;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let button_id = WidgetId::reserved(1);
        let button = EventLogger::new(
            event_sender,
            Button::new("Hello").with_reserved_id(button_id),
        );

        Harness::run_test_window(button, |harness| {
            assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(button_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);
        });
    }
}
