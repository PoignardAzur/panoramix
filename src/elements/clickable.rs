use crate::ctx::ReconcileCtx;
use crate::element_tree::{Element, ElementExt, VirtualDom};
use crate::glue::{Action, GlobalEventCx, WidgetId};
use crate::metadata::{Metadata, NoState};
use crate::widgets::ClickableWidget;

use tracing::{instrument, trace};

// TODO - doc
#[derive(Clone, Debug, Default, PartialEq)]
pub struct Clickable<Child: Element> {
    pub child: Child,
    pub reserved_widget_id: Option<WidgetId>,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ClickableData<Child: VirtualDom> {
    pub child: Child,
    pub reserved_widget_id: Option<WidgetId>,
}

/// Event emitted when a [Clickable] is clicked.
///
/// Note: Might hold data like "mouse position" or "widget id" future versions.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct ClickEvent;

//
// --- IMPLS

impl<Child: Element> Clickable<Child> {
    pub fn new(child: Child) -> Self {
        Clickable {
            child,
            reserved_widget_id: None,
        }
    }

    /// For unit tests only.
    ///
    /// The widget created by this element always has the same id. If two widgets are created
    /// with the same id (for instance, because the same widget is returned twice because of
    /// a copy-paste error), impredictable behavior may ensue.
    pub fn with_reserved_id(self, widget_id: WidgetId) -> Self {
        Clickable {
            reserved_widget_id: Some(widget_id),
            ..self
        }
    }

    /// Provide a closure to be called when the child widget is clicked.
    pub fn on_click<ComponentEvent: 'static, ComponentState: 'static>(
        self,
        md: Metadata<ComponentEvent, ComponentState>,
        callback: impl Fn(&mut ComponentState, ClickEvent) + Clone + 'static,
    ) -> impl Element {
        self.on(md, callback)
    }
}

impl<Child: Element> Element for Clickable<Child> {
    type Event = ClickEvent;

    type ComponentState = NoState;
    type AggregateChildrenState = Child::AggregateChildrenState;
    type BuildOutput = ClickableData<Child::BuildOutput>;

    #[instrument(name = "Clickable", skip(self, prev_state))]
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (
        ClickableData<Child::BuildOutput>,
        Self::AggregateChildrenState,
    ) {
        let (element, child_state) = self.child.build(prev_state);
        (
            ClickableData {
                child: element,
                reserved_widget_id: self.reserved_widget_id,
            },
            child_state,
        )
    }
}

impl<Child: VirtualDom> VirtualDom for ClickableData<Child> {
    type Event = ClickEvent;
    type AggregateChildrenState = Child::AggregateChildrenState;
    type TargetWidgetSeq = ClickableWidget<Child::TargetWidgetSeq>;

    #[instrument(name = "Clickable", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        let id = self.reserved_widget_id.unwrap_or_else(WidgetId::next);
        ClickableWidget::new(self.child.init_tree(), id)
    }

    #[instrument(name = "Clickable", skip(self, prev_value, widget_seq, ctx))]
    fn reconcile(
        &self,
        prev_value: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        self.child
            .reconcile(&prev_value.child, widget_seq.children_mut(), ctx);
    }

    #[instrument(name = "Clickable", skip(self, _children_state, widget, cx))]
    fn process_local_event(
        &self,
        _children_state: &mut Self::AggregateChildrenState,
        widget: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<ClickEvent> {
        // FIXME - Rework event dispatching
        let id = widget.id();
        if let Some(Action::Clicked) = cx.app_data.dequeue_action(id) {
            trace!("Processed click");
            Some(ClickEvent)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::elements::label::Label;
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn new_clickable() {
        let clickable = Clickable::new(Label::new("Hello"));

        let (clickable_data, ()) = clickable.clone().build(());

        assert_debug_snapshot!(clickable);
        assert_debug_snapshot!(clickable_data);
    }

    #[test]
    fn clickable_widget() {
        let clickable = Clickable::new(Label::new("Hello"));

        Harness::run_test_window(clickable, |_harness| {
            // FIXME
            //let clickable_state = harness.get_root_debug_state();
            //assert_debug_snapshot!(clickable_state);

            // FIXME - Test reconcile() method with child
        });
    }

    #[test]
    fn clickable_press() {
        use crate::elements::event_logger::EventLogger;
        use std::sync::mpsc::channel;

        let (event_sender, event_receiver) = channel();
        let clickable_id = WidgetId::reserved(1);
        let clickable = EventLogger::new(
            event_sender,
            Clickable::new(Label::new("Hello")).with_reserved_id(clickable_id),
        );

        Harness::run_test_window(clickable, |harness| {
            // FIXME
            // assert_debug_snapshot!(harness.get_root_debug_state());

            harness.mouse_click_on(clickable_id);

            let click_event = event_receiver.try_recv();
            assert_debug_snapshot!(click_event);
        });
    }
}
