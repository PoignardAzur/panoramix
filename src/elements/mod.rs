//! GUI elements that can be built in a component.

mod any_element;
mod button;
mod checkbox;
mod component;
mod element_list;
mod element_option;
mod empty;
mod flex_element;
mod label;
mod textbox;

pub mod element_tuple;
mod event_logger;
mod mock_component;
mod with_event;

mod compute_diff;

pub use any_element::ElementBox;
pub use button::{Button, ButtonClick};
pub use checkbox::{Checkbox, Toggled};
pub use component::{Component, ComponentOutput};
pub use element_list::ElementList;
pub use element_option::*;
pub use empty::EmptyElement;
pub use flex_element::Flex;
pub use label::Label;
pub use textbox::{TextBox, TextChanged};

// TODO - doc
pub mod internals {
    pub use super::any_element::VirtualDomBox;
    pub use super::button::ButtonData;
    pub use super::checkbox::CheckboxData;
    pub use super::component::{ComponentHolder, ComponentOutputData};
    pub use super::element_list::ElementListData;
    pub use super::element_tuple::ElementTupleData;
    pub use super::empty::EmptyElementData;
    pub use super::flex_element::FlexData;
    pub use super::label::LabelData;
    pub use super::textbox::TextBoxData;
    pub use super::with_event::WithEventTarget;
    pub use super::with_event::{ParentEvent, WithBubbleEvent, WithCallbackEvent, WithMapEvent};

    // TODO - move to test_harness?
    pub use super::event_logger::{EventLogger, EventLoggerData};
    pub use super::mock_component::{MockComponent, MockComponentData, MockState};

    pub use super::compute_diff::{compute_diff, ListMutation, ListMutationItem};
}
