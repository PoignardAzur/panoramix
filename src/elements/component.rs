use crate::ctx::{CompCtx, ProcessEventCtx, ReconcileCtx};
use crate::element_tree::{Element, VirtualDom};
use crate::elements::ElementBox;
use crate::glue::GlobalEventCx;
use crate::metadata::{Metadata, NoState};

use derivative::Derivative;
use std::fmt::Debug;

pub trait Component: Debug + Clone {
    type Props: Clone + Default + Debug + PartialEq + 'static;
    type LocalEvent: Clone + Debug + PartialEq + 'static;

    fn new(props: Self::Props) -> ElementBox<Self::LocalEvent>;

    fn name() -> &'static str;
}

#[derive(Derivative, Default, PartialEq, Eq, Hash)]
#[derivative(Clone(bound = "Comp::Props: Clone"))]
pub struct ComponentHolder<
    Comp: Component,
    ReturnedTree: Element<Event = Comp::LocalEvent>,
    CompFn: Clone + Fn(&CompCtx, Comp::Props) -> ReturnedTree,
> {
    component_fn: CompFn,
    props: Comp::Props,
    _marker: std::marker::PhantomData<Comp>,
}

#[derive(Derivative, Hash)]
#[derivative(
    Debug(bound = ""),
    Clone(bound = ""),
    Default(bound = "Child: Default"),
    PartialEq(bound = "Child: PartialEq"),
    Eq(bound = "Child: Eq")
)]
pub struct ComponentOutput<
    ComponentEvent: Clone + Debug + PartialEq,
    ComponentState: Clone + Default + Debug + PartialEq,
    Child: Element,
> {
    pub child: Child,
    #[derivative(Debug = "ignore")]
    pub _metadata: Metadata<ComponentEvent, ComponentState>,
}

#[derive(Derivative, Hash)]
#[derivative(
    Debug(bound = ""),
    Clone(bound = "Child: Clone"),
    Default(bound = "Child: Default"),
    PartialEq(bound = "Child: PartialEq"),
    Eq(bound = "Child: Eq")
)]
pub struct ComponentOutputData<
    ComponentEvent: Clone + Debug + PartialEq,
    ComponentState: Clone + Default + Debug + PartialEq,
    Child: VirtualDom,
> {
    pub child: Child,
    #[derivative(Debug = "ignore")]
    pub _metadata: Metadata<ComponentEvent, ComponentState>,
}

// ---

impl<
        ComponentEvent: Clone + Debug + PartialEq,
        ComponentState: Clone + Default + Debug + PartialEq,
        Child: Element,
    > ComponentOutput<ComponentEvent, ComponentState, Child>
{
    pub fn new(md: Metadata<ComponentEvent, ComponentState>, child: Child) -> Self {
        Self {
            child,
            _metadata: md,
        }
    }
}

impl<
        Comp: Component,
        ReturnedTree: Element<Event = Comp::LocalEvent>,
        CompFn: Clone + Fn(&CompCtx, Comp::Props) -> ReturnedTree,
    > ComponentHolder<Comp, ReturnedTree, CompFn>
{
    pub fn new(component_fn: CompFn, props: Comp::Props) -> Self {
        Self {
            component_fn,
            props,
            _marker: Default::default(),
        }
    }
}

impl<
        Comp: Component,
        ReturnedTree: Element<Event = Comp::LocalEvent>,
        CompFn: Clone + Fn(&CompCtx, Comp::Props) -> ReturnedTree,
    > std::fmt::Debug for ComponentHolder<Comp, ReturnedTree, CompFn>
{
    #[rustfmt::skip]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple(Comp::name())
            .field(&self.props)
        .finish()
    }
}

// ---

impl<
        Comp: Component + 'static,
        ReturnedTree: Element<Event = Comp::LocalEvent>,
        CompFn: Clone + Fn(&CompCtx, Comp::Props) -> ReturnedTree + 'static,
    > Element for ComponentHolder<Comp, ReturnedTree, CompFn>
{
    type Event = Comp::LocalEvent;
    type ComponentState = NoState;
    type AggregateChildrenState = ReturnedTree::AggregateChildrenState;
    type BuildOutput = ReturnedTree::BuildOutput;

    // TODO - add spans
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState) {
        let default_state = Default::default();
        let local_state = ReturnedTree::get_component_state(&prev_state).unwrap_or(&default_state);

        let ctx = CompCtx {
            called_use_metadata: std::cell::Cell::new(false),
            local_state: local_state,
        };
        let element_tree = (self.component_fn)(&ctx, self.props);

        element_tree.build(prev_state)
    }
}

/// ---

impl<
        ComponentEvent: Clone + Debug + PartialEq + 'static,
        ComponentState: Clone + Default + Debug + PartialEq + 'static,
        Child: Element,
    > Element for ComponentOutput<ComponentEvent, ComponentState, Child>
{
    type Event = ComponentEvent;

    type ComponentState = ComponentState;
    // TODO - Store Event queue somewhere else?
    type AggregateChildrenState = (
        Vec<ComponentEvent>,
        ComponentState,
        Child::AggregateChildrenState,
    );
    type BuildOutput = ComponentOutputData<ComponentEvent, ComponentState, Child::BuildOutput>;

    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState) {
        let (_, prev_local_state, children_prev_state) = prev_state;
        let (child, children_state) = self.child.build(children_prev_state);
        (
            ComponentOutputData {
                child,
                _metadata: Default::default(),
            },
            (vec![], prev_local_state, children_state),
        )
    }

    fn get_component_state(state: &Self::AggregateChildrenState) -> Option<&Self::ComponentState> {
        Some(&state.1)
    }
}

impl<
        ComponentEvent: Clone + Debug + PartialEq + 'static,
        ComponentState: Clone + Default + Debug + PartialEq + 'static,
        Child: VirtualDom,
    > VirtualDom for ComponentOutputData<ComponentEvent, ComponentState, Child>
{
    type Event = ComponentEvent;
    type AggregateChildrenState = (
        Vec<ComponentEvent>,
        ComponentState,
        Child::AggregateChildrenState,
    );
    type TargetWidgetSeq = Child::TargetWidgetSeq;

    // TODO - add spans
    fn init_tree(&self) -> Child::TargetWidgetSeq {
        self.child.init_tree()
    }

    fn reconcile(
        &self,
        prev_value: &Self,
        widget_seq: &mut Child::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        self.child.reconcile(&prev_value.child, widget_seq, ctx);
    }

    fn process_local_event(
        &self,
        children_state: &mut Self::AggregateChildrenState,
        _widget_seq: &mut Child::TargetWidgetSeq,
        _cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        let event_queue = &mut children_state.0;
        // TODO - this is a stack, not a queue; whatever, I'll use VecDeque later
        event_queue.pop()
    }

    fn process_event(
        &self,
        _comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) {
        let mut ctx = ProcessEventCtx {
            event_queue: &mut children_state.0,
            state: &mut children_state.1,
        };
        self.child
            .process_event(&mut ctx, &mut children_state.2, widget_seq, cx)
    }
}

#[cfg(test)]
mod tests {
    #![allow(dead_code)]

    use crate as panoramix;

    #[derive(Debug, Default, Clone, PartialEq, Hash)]
    struct MyComponent;

    type MyPropsType = ();
    type MyLocalEvent = panoramix::NoEvent;

    impl MyComponent {
        fn new(props: MyPropsType) -> impl panoramix::Element<Event = MyLocalEvent> {
            <Self as panoramix::elements::component::Component>::new(props)
        }

        fn render(
            _ctx: &panoramix::CompCtx,
            _my_props: MyPropsType,
        ) -> impl panoramix::Element<Event = MyLocalEvent> {
            panoramix::elements::EmptyElement::new()
        }
    }

    impl panoramix::elements::component::Component for MyComponent {
        type Props = MyPropsType;
        type LocalEvent = MyLocalEvent;

        fn new(props: Self::Props) -> panoramix::elements::ElementBox<MyLocalEvent> {
            panoramix::elements::ElementBox::new(panoramix::elements::internals::ComponentHolder::<
                Self,
                _,
                _,
            >::new(&MyComponent::render, props))
        }

        fn name() -> &'static str {
            "MyComponent"
        }
    }

    use crate::element_tree::Element;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn call_component() {
        let my_component = MyComponent::new(());

        let (component_result, _state) = my_component.build(Default::default());
        assert_debug_snapshot!(component_result);

        //let prev_state = (999, Default::default());
        //let (component_result, component_state) = my_component.build(prev_state);
        //assert_eq!(component_state.0, 999);

        // TODO - local state
        // TODO - process_event
    }

    // TODO
    // - Widget test
    // - Events
}
