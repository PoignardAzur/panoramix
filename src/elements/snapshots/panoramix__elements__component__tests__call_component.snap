---
source: src/elements/component.rs
expression: component_result
---
MyComponent(
    EmptyElementData,
)
